File: src/Monet/MOButton.m
#import <Monet/MOButton.h>

#import <Monet/MOApplication.h>
#import <Monet/MOPoint.h>
#import <Monet/MORect.h>
#import <Monet/MOView.h>

struct MOButtonData
{
	BOOL isMouseDown;
	BOOL isMouseInside;
};

@implementation MOButton : MOView

- (id)initWithFrame: (MORect)aFrame app: (MOApplication *)aApp;
{
	if ((self = [super initWithFrame: aFrame app: aApp]))
	{
		buttonData = calloc(1, sizeof (struct MOButtonData));
		buttonData->isMouseDown   = NO;
		buttonData->isMouseInside = NO;
	}

	return self;
}

#pragma mark -

- (BOOL)isPressed
{
	return buttonData->isMouseDown && MORectContainsPoint(
		[self bounds],
		[self convertPointFromScreen: [[self app] mouseLocation]]);
}

#pragma mark -

- (void)mouseDown: (MOEvent *)aEvent
{
	buttonData->isMouseDown = YES;
}

- (void)mouseUp: (MOEvent *)aEvent
{
	if ([self isPressed])
		[self clicked];

	buttonData->isMouseDown = NO;
}

#pragma mark -

- (void)clicked
{
	;
}

@end

File: src/Monet/MOButton.c
#include <Monet/MOButton.h>

#include <Monet/Common.h>

#include <Monet/MOApplication.h>
#include <Monet/MOPoint.h>
#include <Monet/MORect.h>
#include <Monet/MOView.h>

COClass MOButtonClass = {
	.superclass = &MOViewClass,
	.destructor = NULL
};

bool _MOButtonIsMousePressed(MOButton *aButton)
{
	MOView *view = (MOView *)aButton;

	MORect bounds = MOViewGetBounds(view);
	MOPoint untranslatedPoint = MOApplicationGetMouseLocation(MOViewGetApplication(view));
	MOPoint point = MOViewConvertPointFromScreen(view, untranslatedPoint);

	return aButton->isMouseDown && MORectContainsPoint(bounds, point);
}

bool _MOButtonMouseButtonPressed(MOView *aView, MOEvent *aEvent)
{
	MOButton *button = (MOButton *)aView;

	button->isMouseDown = true;

	return true;
}

bool _MOButtonMouseButtonReleased(MOView *aView, MOEvent *aEvent)
{
	MOButton *button = (MOButton *)aView;

	if (_MOButtonIsMousePressed(button))
		button->clickedCallback(button);

	button->isMouseDown = false;

	return true;
}

void MOButtonInit(MOButton *aButton, MORect aFrame, MOApplication *aApplication)
{
	MOView *view = MOButtonAsView(aButton);

	MOViewInit(view, aFrame, aApplication);

	MOViewSetMouseButtonPressedCallback(view, &_MOButtonMouseButtonPressed);
	MOViewSetMouseButtonReleasedCallback(view, &_MOButtonMouseButtonReleased);
}

MOView *MOButtonAsView(MOButton *aButton)
{
	return (MOView *)aButton;
}

void MOButtonSetClickedCallback(MOButton *aButton, MOButtonClickedCallback aCallback)
{
	aButton->clickedCallback = aCallback;
}
